File: src/main/java/AoC2023_13.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_13
        extends SolutionBase<List<CharGrid>, Integer, Integer> {
    
    private AoC2023_13(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_13 create() {
        return new AoC2023_13(false);
    }
    
    public static AoC2023_13 createDebug() {
        return new AoC2023_13(true);
    }
    
    @Override
    protected List<CharGrid> parseInput(final List<String> inputs) {
        return StringOps.toBlocks(inputs).stream().map(CharGrid::from).toList();
    }
    
    private Symmetry findSymmetry(final CharGrid grid, final int smudge) {
        return Stream.concat(
            range(grid.getWidth() - 1).intStream()
                .mapToObj(col -> {
                    final int errors = countHorizontalSymmetryErrors(grid, col);
                    return new Symmetry(0, col + 1, errors);
                })
                .filter(s -> s.errors() == smudge),
            range(grid.getHeight() - 1).intStream()
                .mapToObj(row -> {
                    final int errors = countVerticalSymmetryErrors(grid, row);
                    return new Symmetry(row + 1, 0, errors);
                })
                .filter(s -> s.errors() == smudge)
        ).findFirst().orElseThrow();
    }

    private int countVerticalSymmetryErrors(final CharGrid grid, final int row) {
        return range(grid.getHeight() / 2).intStream()
            .map(dr -> {
                final int up = row - dr;
                final int down = row + dr + 1;
                if (!(0 <= up && up < down && down < grid.getHeight())) {
                    return 0;
                }
                return (int) range(grid.getWidth()).intStream()
                    .filter(c -> grid.getValue(Cell.at(up, c)) != grid.getValue(Cell.at(down, c)))
                    .count();
            })
            .sum();
    }

    private int countHorizontalSymmetryErrors(final CharGrid grid, final int col) {
        return range(grid.getWidth() / 2).intStream()
            .map(dc -> {
                final int left = col - dc;
                final int right = col + dc + 1;
                if (!(0 <= left && left < right && right < grid.getWidth())) {
                    return 0;
                }
                return (int) range(grid.getHeight()).intStream()
                        .filter(r -> grid.getValue(Cell.at(r, left)) != grid.getValue(Cell.at(r, right)))
                        .count();
            })
            .sum();
    }
    
    private int solve(final List<CharGrid> grids, final int smudge) {
        return grids.stream()
            .map(g -> findSymmetry(g, smudge))
            .mapToInt(s -> 100 * s.row() + s.column())
            .sum();
    }

    @Override
    public Integer solvePart1(final List<CharGrid> grids) {
        return solve(grids, 0);
    }
    
    @Override
    public Integer solvePart2(final List<CharGrid> grids) {
        return solve(grids, 1);
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "405"),
        @Sample(method = "part2", input = TEST, expected = "400"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_13.create().run();
    }
    
    record Symmetry(int row, int column, int errors) {}

    private static final String TEST = """
            #.##..##.
            ..#.##.#.
            ##......#
            ##......#
            ..#.##.#.
            ..##..##.
            #.#.##.#.

            #...##..#
            #....#..#
            ..##..###
            #####.##.
            #####.##.
            ..##..###
            #....#..#
            """;
}

File: src/main/python/AoC2023_13.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 13
#

import sys

from aoc import my_aocd
from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import CharGrid

Input = list[CharGrid]
Output1 = int
Output2 = int


TEST = """\
#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [
            CharGrid.from_strings(block)
            for block in my_aocd.to_blocks(input_data)
        ]

    def solve(self, grids: list[CharGrid], smudge: int) -> int:
        def find_symmetry(grid: CharGrid, smudge: int) -> tuple[int, int]:
            h = grid.get_height()
            w = grid.get_width()
            for c in range(w - 1):
                err = 0
                for dc in range(w // 2):
                    left = c - dc
                    right = c + dc + 1
                    if not 0 <= left < right < w:
                        continue
                    err += sum(
                        1
                        for r in range(h)
                        if grid.values[r][left] != grid.values[r][right]
                    )
                if err == smudge:
                    return (0, c + 1)
            for r in range(h - 1):
                err = 0
                for dr in range(h // 2):
                    up = r - dr
                    down = r + dr + 1
                    if not 0 <= up < down < h:
                        continue
                    err += sum(
                        1
                        for c in range(w)
                        if grid.values[up][c] != grid.values[down][c]
                    )
                if err == smudge:
                    return (r + 1, 0)
            raise AssertionError

        return sum(
            (100 * r) + c
            for r, c in (find_symmetry(grid, smudge) for grid in grids)
        )

    def part_1(self, grids: Input) -> Output1:
        return self.solve(grids, smudge=0)

    def part_2(self, grids: Input) -> Output2:
        return self.solve(grids, smudge=1)

    @aoc_samples(
        (
            ("part_1", TEST, 405),
            ("part_2", TEST, 400),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 13)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
